cfme/base/credential.py

```diff
@@ -0,0 +1,195 @@
+        secret: Something
+        verify_secret: Something
+    """
+    pretty_attrs = ['principal', 'secret']
+
+    def __init__(self, principal, secret, verify_secret=None, domain=None):
+        self.principal = principal
+        self.secret = secret
+        self.verify_secret = verify_secret
+        self.domain = domain
+
+    def __getattribute__(self, attr):
+        if attr == 'verify_secret':
+            if object.__getattribute__(self, 'verify_secret') is None:
+                return object.__getattribute__(self, 'secret')
+            else:
+                return object.__getattribute__(self, 'verify_secret')
+
+        elif attr == 'principal':
+            domain = object.__getattribute__(self, 'domain')
+            principal = object.__getattribute__(self, 'principal')
+            return r'{}\{}'.format(domain, principal) if domain else principal
+        else:
+            return super(Credential, self).__getattribute__(attr)
+
+    def as_fill_value(self):
+        """
+        used for filling forms like add/edit provider form
+        Returns: dict
+        """
+        return {
+            'username': self.principal,
+            'password': self.secret,
+            'confirm_password': self.verify_secret
+        }
+
+    @property
+    def form(self):
+        return provider_credential_form()
+
+
+class AMQPCredential(Credential):
+    pass
+
+
+class CANDUCredential(Credential):
+    pass
+
+
+class AzureCredential(Credential):
+    pass
+
+
+class SSHCredential(Credential):
+    pass
+
+
+class TokenCredential(Pretty, Updateable, FromConfigMixin):
+    """
+    A class to fill in credentials
+
+    Args:
+        token: Something
+        verify_token: Something
+    """
+    pretty_attrs = ['token']
+
+    def __init__(self, token, verify_token=None):
+        self.token = token
+        self.verify_token = verify_token
+
+    def __getattribute__(self, attr):
+        if attr == 'verify_token':
+            if object.__getattribute__(self, 'verify_token') is not None:
+                return object.__getattribute__(self, 'verify_token')
+            else:
+                return object.__getattribute__(self, 'token')
+        else:
+            return super(TokenCredential, self).__getattribute__(attr)
+
+    def as_fill_value(self):
+        """
+        used for filling forms like add/edit provider form
+        Returns: dict
+        """
+        return {
+            'token': self.token,
+            'verify_token': self.verify_token
+        }
+
+    @property
+    def form(self):
+        return provider_credential_form()
+
+
+class ServiceAccountCredential(Pretty, Updateable, FromConfigMixin):
+    """
+    A class to fill in credentials
+
+    Args:
+        service_account: Something
+    """
+    pretty_attrs = ['service_account']
+
+    def __init__(self, service_account):
+        super(ServiceAccountCredential, self)
+        self.service_account = service_account
+
+    def as_fill_value(self):
```

Comment:
MIssing Fillable inheritance on this class, make sure all the credentials classes have it.